zip-0227.rst

```diff
@@ -1,8 +1,407 @@
+
+For the `IssueBundle`:
+
+- encode the `vIssueActions` vector
+- encode the :math:`\mathsf{ik}` as 32 byte-string
+- sign the `SIGHASH` of the transaction with the issuance authorizing key, :math:`\mathsf{isk}`, using the RedPallas signature scheme. The signature is then added to the issuance bundle.
+
+
+NOTE that the commitment is not included in the `IssuanceAction` itself. As explained below, it is later computed by the validators and added to the `NoteCommitmentTree`.
+
+Concrete Applications
+---------------------
+
+**Bridging Assets**
+Issuers can wrap Assets defined in other chains and issue them at once in a single transaction. The specifics of the bridge are not implemented in the protocol, but there are several ways we conceive issuers can build these bridges (at least centralized bridges):
+
+- First, the issuance and burn mechanism can be used in conjunction to determine the 
+
+**Asset Features**
+
+- By using the ``finalize`` boolean and the burning mechanism defined in [#zip-0226]_, issuers can control the supply production of any Asset associated to their issuer keys. For example,
+
+    - by setting ``finalize = 1`` from the first issuance action for that Asset Identifier, the issuer is in essence creating a one-time issuance transaction. This is useful when the max supply is capped from the beginning and the distribution is known in advance. All tokens are issued at once and distributed as needed.
+
+- Issuers can also stop the existing supply production of any Asset associated to their issuer keys. This could be done by
+
+    - issuing a last set of tokens of that specific :math:`\mathsf{AssetId}`, during which ``finalize = 1``, or by
+    - issuing a transaction with a single note in the issuance action pertaining to that :math:`\mathsf{AssetId}`, where the note will contain a ``value = 0``. This can be used for application-specific purposes (NFT collections) or for security purposes to revoke the Asset issuance (see Security and Privacy Considerations).
+
+- Furthermore, NFT issuance is enabled by issuing in a single bundle several issuance actions, where each :math:`\mathsf{AssetId}` corresponds to ``value = 1`` at the fundamental unit level. Issuers and users should make sure that ``finalize = 1`` for each of the actions in this scenario.
+
+Consensus Rule Changes
+----------------------
+
+For the IssueBundle:
+
+- Verify the RedPallas-based issuance authorization signature on `SIGHASH`, `authorization`, is verified by invoking `authorization.VerifySig(ik, SIGHASH)`
+
+For each `IssueAction` in `IssueBundle`:
+
+- check that :math:`\mathsf{asset\_desc}` is a string with `0 < byte size <= 512`.
+- retrieve :math:`\mathsf{AssetId}` from the first note in the sequence and check that :math:`\mathsf{AssetId = GroupHash^{\mathbb{P}}(ik || asset\_desc)}` is properly derived.
+- check that the :math:`\mathsf{AssetId}` does not exist in the ``previously_finalized`` set in the global state.
```

Comment:
This rule is ambiguous, there are two possible interpretations:
1. `finalize` must not have been set to `1` in any previous _transaction_ in this chain
2. `finalize` must not have been set to `1` in any previous _block_ in this chain

Alternative 1 is the most consistent with the transparent rules, which allow spends of outputs created in the same block.

Alternative 2 is the most consistent with the shielded rules, where spends require an anchor from a previous block. It also makes parallel verification and block production easier to implement, because we don't have to check an extra rule for transactions in the same block.

Reply:
Note the interaction between parallel verification and whether it is possible to say that an issuance tx with `finalize = 1` is the "last transaction" that issues the corresponding asset. If no interaction between issuance txs in a block is desired then we need Alternative 2 *and also* no check that there is not more than one tx with `finalize = 1` in a block for the same asset. In that case, describing a tx with `finalize = 1` is the last issuance tx for the asset would be incorrect.

Reply:
We updated to use Alternative 2.